core/src/accounts_hash_verifier.rs

```diff
@@ -155,17 +179,68 @@ impl AccountsHashVerifier {
             }
         }
 
+        cluster_info.push_accounts_hashes(hashes.clone());
+    }
+
+    fn submit_for_packaging(
+        accounts_package: AccountsPackage,
+        pending_snapshot_package: Option<&PendingSnapshotPackage>,
+        snapshot_config: Option<&SnapshotConfig>,
+    ) {
         if let Some(snapshot_config) = snapshot_config {
-            if snapshot_package.block_height % snapshot_config.full_snapshot_archive_interval_slots
-                == 0
-            {
-                if let Some(pending_snapshot_package) = pending_snapshot_package {
+            if let Some(pending_snapshot_package) = pending_snapshot_package {
+                if accounts_package.block_height
+                    % snapshot_config.full_snapshot_archive_interval_slots
+                    == 0
+                {
+                    let snapshot_package = SnapshotPackage::new(
+                        accounts_package,
+                        snapshot_config.snapshot_package_output_path.clone(),
+                        snapshot_config.archive_format,
+                        None,
+                    );
+
                     *pending_snapshot_package.lock().unwrap() = Some(snapshot_package);
-                }
-            }
-        }
+                } else if accounts_package.block_height
+                    % snapshot_config.incremental_snapshot_archive_interval_slots
```

Comment:
I am not sure how to configure when the snapshots is taken. I think it would be simpler to say that we take a snapshot for every X amount of intervals. And among these we take a full snapshot for every Y snapshots.
The double intervals based on the block_height is difficult to understand -- at least to me.

Reply:
Yeah, this took me a bit to understand too. There are three intervals at play, and two "slots". First the intervals:

- accounts hash interval
- incremental snapshot interval
- full snapshot interval

The accounts hash interval is the interval used to take a bank snapshot, and also what gets pushed to AccountsHashVerifier and out to the cluster.

The full and incremental snapshot intervals each just say when to take their respective snapshots. Based on the code, full and incremental snapshots _must_ be a multiple of the accounts hash interval (otherwise the modulus may never hit zero). This is enforced in the CLI parameters validation.

One additional thing is that full snapshots take precedence over incremental snapshots. I.e. if the current slot ends up being an interval for _both_ full and incremental snapshots, then we only take a full snapshot.

And that does it for the intervals. Next are the two slots: the "slot" and the "block height".

You'll see that block height is used everywhere when comparing to the intervals. This is because it's possible for a slot number to just not exist. I.e. slots could validly be 1, 2, 3, 4, 6, 7, 8, 12, 13, 14, 17. If one of the intervals was 5, it would never get hit. Thus, block height. The block height never has gaps, so the interval checks never get skipped.

Does that make sense? Let me know if there are further questions. I'd love to make this part simpler if possible too!